hotshot-state-prover/src/service.rs

```diff
@@ -423,6 +453,148 @@ pub async fn sync_state<ApiVer: StaticVersionType>(
     let proof_gen_elapsed = Instant::now().signed_duration_since(proof_gen_start);
     tracing::info!("Proof generation completed. Elapsed: {proof_gen_elapsed:.3}");
 
+    Ok((proof, public_input))
+}
+
+async fn epoch_catchup(
+    state: &mut ProverServiceState,
+    provider: &impl Provider,
+    light_client_address: Address,
+    mut cur_st_state: StakeTableState,
+    proving_key: &ProvingKey,
+    contract_epoch: u64,
+    target_epoch: u64,
+) -> Result<(), ProverError> {
+    for epoch in contract_epoch..target_epoch {
+        tracing::info!("Advancing to epoch {}...", epoch + 1);
+        state
+            .sync_with_epoch(epoch)
+            .await
+            .map_err(ProverError::NetworkError)?;
+        let state_cert =
+            fetch_epoch_state_from_sequencer(&state.config.sequencer_url, epoch).await?;
+        let signature_map = state_cert
+            .signatures
+            .into_iter()
+            .collect::<HashMap<StateVerKey, StateSignature>>();
+
+        let (proof, public_input) = generate_proof_helper(
+            state,
+            &state_cert.light_client_state,
+            &cur_st_state,
+            &state_cert.next_stake_table_state,
+            signature_map,
+            proving_key,
+        )
+        .await?;
+
+        submit_state_and_proof(provider, light_client_address, proof, public_input).await?;
```

Comment:
i think what's particularly confusing is that we have both the `epoch_catchup` and `sync_state` to generate the proof and submit to contract.

i find it little hard to reason about its correctness. It'll be great if we can simplify and modularize/reorg the logic somehow that we only have `generate_proof` and `submit_proof` once.

Reply:
I'm not sure if I understand this correctly. You cannot aggregate proofs of multiple epochs